src/hermes.rs

```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    pub fn from_slice(slice: &[u8]) -> Result<Self> {
+        match decode_slice(slice)? {
+            DecodedMap::Hermes(sm) => Ok(sm),
+            _ => Err(Error::IndexedSourcemap),
+        }
+    }
+
+    pub fn to_writer<W: Write>(&self, w: W) -> Result<()> {
+        encode(self, w)
+    }
+
+    pub fn get_original_function_name(&self, bytecode_offset: u32) -> Option<&str> {
+        let token = self.sm.lookup_token(0, bytecode_offset)?;
+
+        let function_map = &self.function_maps[token.get_src_id() as usize].as_ref()?;
+
+        // Find the closest mapping, just like here:
+        // https://github.com/facebook/metro/blob/63b523eb20e7bdf62018aeaf195bb5a3a1a67f36/packages/metro-symbolicate/src/SourceMetadataMapConsumer.js#L204-L231
+        let mapping =
+            function_map
+                .mappings
+                .binary_search_by(|o| match o.line.cmp(&token.get_src_line()) {
+                    Ordering::Equal => o.column.cmp(&token.get_src_col()),
+                    x => x,
+                });
+        let name_index = function_map.mappings[match mapping {
+            Ok(a) => a,
+            Err(a) => a.saturating_sub(1),
+        }]
+        .name_index;
+
+        function_map
+            .names
+            .get(name_index as usize)
+            .map(|n| n.as_str())
+    }
+
+    pub fn rewrite(self, options: &RewriteOptions<'_>) -> Result<Self> {
+        let Self {
+            sm,
+            function_maps,
+            raw_facebook_sources,
+        } = self;
+        let sm = sm.rewrite(options)?;
+        Ok(Self {
+            sm,
+            function_maps,
+            raw_facebook_sources,
+        })
+    }
+}
+
+pub fn decode_hermes(mut rsm: RawSourceMap) -> Result<SourceMapHermes> {
+    let x_facebook_sources = rsm
+        .x_facebook_sources
+        .take()
+        .expect("expected x_facebook_sources");
+
+    // This is basically the logic from here:
+    // https://github.com/facebook/metro/blob/63b523eb20e7bdf62018aeaf195bb5a3a1a67f36/packages/metro-symbolicate/src/SourceMetadataMapConsumer.js#L182-L202
+
+    let function_maps = x_facebook_sources
+        .iter()
+        .map(|v| {
+            let FacebookScopeMapping {
+                names,
+                mappings: raw_mappings,
+            } = v.as_ref()?.iter().next()?;
+
+            let mut mappings = vec![];
+            let mut line = 1;
+            let mut name_index = 0;
+
+            for line_mapping in raw_mappings.split(';') {
+                if line_mapping.is_empty() {
+                    continue;
+                }
+
+                let mut column = 0;
+
+                for mapping in line_mapping.split(',') {
+                    if mapping.is_empty() {
+                        continue;
+                    }
+
+                    let mut nums = parse_vlq_segment(mapping).ok()?.into_iter();
+
+                    column = (i64::from(column) + nums.next()?) as u32;
```

Comment:
wonder if we want to deal with the overflow explicitly here.

Reply:
hm, the regular sourcemap code doesn’t either.